File: ESTRUTURAS/3171/main.dart
import 'dart:io';
import 'dart:collection';

bool canFix(int l) {
  Queue<List<int>> segments = new Queue();
  for (int times = 0; times < l; times++) {
    segments.add(stdin.readLineSync()!.split(' ').map(int.parse).toList());
  }
  Set connected = segments.removeFirst().toSet();
  bool hasChange = true;
  while (hasChange) {
    hasChange = false;
    for (int times = 0; times < segments.length; times++) {
      var [x, y] = segments.removeFirst();
      if (connected.contains(x) || connected.contains(y)) {
        connected.add(x);
        connected.add(y);
        hasChange = true;
      } else {
        segments.addLast([x, y]);
      }
    }
  }
  return segments.isEmpty;
}

void main() {
  int l = int.parse(stdin.readLineSync()!.split(' ')[1]);
  print(canFix(l) ? 'COMPLETO' : 'INCOMPLETO');
}

File: ESTRUTURAS/3171/main.rb
# frozen_string_literal:true

require 'set'

def can_fix(l)
  segments = []
  l.times do
    segments.push(gets.split.map(&:to_i))
  end
  connected = segments.shift.to_set
  has_change = true
  while has_change
    has_change = false
    segments.size.times do
      x, y = segments.shift
      if (connected.include? x) || (connected.include? y)
        connected.merge([x, y])
        has_change = true
      else
        segments.push([x, y])
      end
    end
  end
  segments.empty?
end

l = gets.split[1].to_i
puts can_fix(l) ? 'COMPLETO' : 'INCOMPLETO'
